lightningd/pay.c

```diff
@@ -625,62 +634,76 @@ bool wait_payment(const tal_t *cxt,
 	return cb_not_called;
 }
 
-/* Returns false if cb was called, true if cb not yet called. */
-bool
-send_payment(const tal_t *ctx,
-	     struct lightningd* ld,
-	     const struct sha256 *rhash,
-	     const struct route_hop *route,
-	     u64 msatoshi,
-	     void (*cb)(const struct sendpay_result *, void*),
-	     void *cbarg)
+static struct custom_route_data route_to_custom_route_data(
+	struct lightningd* ld,
+	const struct route_hop *route)
 {
-	const u8 *onion;
-	u8 sessionkey[32];
-	unsigned int base_expiry;
-	struct onionpacket *packet;
-	struct secret *path_secrets;
-	enum onion_type failcode;
 	size_t i, n_hops = tal_count(route);
-	struct hop_data *hop_data = tal_arr(tmpctx, struct hop_data, n_hops);
-	struct pubkey *ids = tal_arr(tmpctx, struct pubkey, n_hops);
-	struct wallet_payment *payment = NULL;
-	struct htlc_out *hout;
-	struct short_channel_id *channels;
-	struct routing_failure *fail;
-	struct channel *channel;
-	struct sendpay_result *result;
+	struct custom_route_data ret;
+	ret.hop_data = tal_arr(tmpctx, struct hop_data, n_hops);
+	ret.ids = tal_arr(tmpctx, struct pubkey, n_hops);
+	ret.channels = tal_arr(tmpctx, struct short_channel_id, n_hops);
 
 	/* Expiry for HTLCs is absolute.  And add one to give some margin. */
-	base_expiry = get_block_height(ld->topology) + 1;
+	ret.base_expiry = get_block_height(ld->topology) + 1;
 
 	/* Extract IDs for each hop: create_onionpacket wants array. */
 	for (i = 0; i < n_hops; i++)
-		ids[i] = route[i].nodeid;
+		ret.ids[i] = route[i].nodeid;
 
-	/* Copy hop_data[n] from route[n+1] (ie. where it goes next) */
+	/* Copy hd[n] from route[n+1] (ie. where it goes next) */
 	for (i = 0; i < n_hops - 1; i++) {
-		hop_data[i].realm = 0;
+		ret.hop_data[i].realm = 0;
 		serialize_per_hop_data(
 			&route[i+1].channel_id,
 			route[i+1].amount,
-			base_expiry + route[i+1].delay,
-			hop_data[i].per_hop_data
+			ret.base_expiry + route[i+1].delay,
+			ret.hop_data[i].per_hop_data
 			);
 	}
 
 	/* And finally set the final hop to the special values in
 	 * BOLT04 */
-	hop_data[i].realm = 0;
+	ret.hop_data[i].realm = 0;
 	struct short_channel_id channel_id;
 	memset(&channel_id, 0, sizeof(channel_id));
 	serialize_per_hop_data(
 		&channel_id,
 		route[i].amount,
-		base_expiry + route[i].delay,
-		hop_data[i].per_hop_data
+		ret.base_expiry + route[i].delay,
+		ret.hop_data[i].per_hop_data
 		);
 
+	/* Copy channels used along the route. */
+	for (i = 0; i < n_hops; ++i)
+		ret.channels[i] = route[i].channel_id;
+
+	return ret;
+}
+
+/* Returns false if cb was called, true if cb not yet called. */
+static bool
+send_payment_custom_route_data(const tal_t *ctx,
```

Comment:
The name is a bit misleading imho, it should be the common entrypoint for injected and locally originating payments.

Reply:
Notice also that this function does quite a lot of book-keeping that should really only be done for payments that originated here. If we later inject an onion from outside we likely don't want it to show up in the `listpayments` output. These should be extracted in a wrapper instead, that is bypassed when injecting.

Reply:
I think, for now, we can see this as a locally originating payment. The extended functionality is not related to an app-specific -> standard-lightning bridge on this node, but to enable a standard-lightning -> app-specific bridge on another node, downstream the route.

The app-specific -> standard-lightning bridge idea is an interesting line of thought. Maybe it should have its own RPC command, or maybe an option in sendpay to consider this routing of someone else's transaction instead of considering this our own transaction. For simplicity, let's not include that in this pull request.